File: dotnet/cs/src/Link/Log.cs
/*
 * File: Log.cs
 * Project: Link
 * Created Date: 13/09/2023
 * Author: Shun Suzuki
 * -----
 * Last Modified: 13/09/2023
 * Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
 * -----
 * Copyright (c) 2023 Shun Suzuki. All rights reserved.
 * 
 */

using System.Runtime.InteropServices;
using AUTD3Sharp.Internal;

namespace AUTD3Sharp.Link
{
    /// <summary>
    /// Link for logging
    /// </summary>
    public sealed class Log : Internal.Link
    {
        public Log(Internal.Link link) : base(NativeMethods.Base.AUTDLinkLog(link.Ptr))
        {
        }

        /// <summary>
        /// Set log function
        /// </summary>
        /// <remarks>By default, the logger will display log messages on the console.</remarks>
        /// <param name="output">output callback</param>
        /// <param name="flush">flush callback</param>
        /// <returns></returns>
        public Log WithLogFunc(OnLogOutputCallback output, OnLogFlushCallback flush)
        {
            Ptr = NativeMethods.Base.AUTDLinkLogWithLogFunc(Ptr, Marshal.GetFunctionPointerForDelegate(output), Marshal.GetFunctionPointerForDelegate(flush));
            return this;
        }

        /// <summary>
        /// Set log level
        /// </summary>
        /// <param name="level"></param>
        /// <returns></returns>
        public Log WithLogLevel(Level level)
        {
            Ptr = NativeMethods.Base.AUTDLinkLogWithLogLevel(Ptr, level);
            return this;
        }
    }

    public static class LogLinkExtensions
    {
        public static Log WithLog(this Internal.Link s)
        {
            return new Log(s);
        }
    }
}

File: cpp/tests/link/log.cpp
// File: log.cpp
// Project: link
// Created Date: 26/09/2023
// Author: Shun Suzuki
// -----
// Last Modified: 27/09/2023
// Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
// -----
// Copyright (c) 2023 Shun Suzuki. All rights reserved.
//

#include <gtest/gtest.h>

#include <autd3/internal/controller.hpp>
#include <autd3/link/debug.hpp>

void test_log_out(const char* msg) { std::cerr << msg; }
void test_log_flush() {}

TEST(Link, Log) {
  const auto autd = autd3::internal::Controller::builder()
                        .add_device(autd3::internal::AUTD3(autd3::internal::Vector3::Zero(), autd3::internal::Vector3::Zero()))
                        .open_with(autd3::link::Debug()
                                       .with_log()
                                       .with_log_func(&test_log_out, &test_log_flush)
                                       .with_log_level(autd3::internal::native_methods::Level::Trace));

  autd.close();
}
